src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -627,6 +649,14 @@ Shard::finalize(
             std::string("exception ") + e.what() + " in function " + __func__);
     }
 
+    if (dsh)
+    {
+        backend_->close();
+        dsh->close();
+        backend_->open(false);
+        return finalize(false, expectedHash, false);
+    }
```

Comment:
Could you please explain the rational of this block of code?

Reply:
backend_->close(); // Close non-deterministic shard database.
dsh->close(); // Replace non-deterministic shard by deterministic one.
backend_->open(false); // Re-open deterministic shard database.

The finalize() function verifies the shard and, if third parameter is true, then replaces the shard by deterministic copy of the shard. After deterministic shard is created it verifies again, The finalize() function called here to verify deterministic shard, third parameter is false.

Done: comment added to code

Reply:
Thanks so much!